### PR TITLE
Adapt parsing to work with certain qif files seen in the wild

### PR DESCRIPTION
**Pull request by a-bird-of-paradise**
Using current release version of `Quiffen` with my bank's proffered `QIF` files resulted in parse errors to do with the date. I used sed to force the dates into a parsable format and then some transactions were dropped. 

I traced these issues to two things. 

- Dates: bank's `QIF` has dates in a `YYYY-MM-DD` format that `Quiffen` could not handle. 
- Transactions: there are some circumstances where the first caret-delimited section has transaction details which are skipped over. 

To correct the date issue, I added `YMD` date patterns that are used as a fallback if the `DMY` or `MDY` patterns don't match. 

For the transaction issue, I added in a transaction parse step when the default account is created. 

Finally, I added a test `QIF` (two real anonymised transactions) and a test case to ensure it is parsed as expected without my changes breaking any existing tested functionality.

### CODE REVIEW

File: quiffen/core/qif.py [hunk 1]
```diff
@@ -191,6 +191,17 @@ def parse(cls, filepath, separator='\n', day_first=True):
                                                'present')
                 accounts[default_account.name] = default_account
                 last_account = default_account.name
+
+                # if there is a transaction in this section, parse it (often seen at head of qifs)
+
+                if len(section.split('\n')) > 2: # !Type:... and ^
```
Comment (isaacharrisholt) [inline quiffen/core/qif.py:197]: Is it only `>2` in the case of a transaction, or are there potentially other things that can appear here?

Comment (a-bird-of-paradise) [reply quiffen/core/qif.py:197]: I think on reflection I'd want to do something a bit different here. 

The issue is that if the first section has a header line and transaction details in it, then a default account is created without processing the transaction details. Per the qif documentation on wikipedia (I know...) there is no caret between the header and the first transaction. So what we should probably do is:

* swap the order of the `elif '!Type:Invst' in header_line:` and `elif '!Type' in header_line and not accounts:` tests
* amend `elif '!Type' in header_line and not accounts:` to `elif not accounts and header_line in VALID_TRANSACTION_ACCOUNT_TYPES:`
* after creating the default account, send the first section to the transaction processor without testing for >2 lines

The investment type already creates the transaction for the first section, it is just the other `VALID_TRANSACTION_ACCOUNT_TYPES` that need the same treatment, assuming Wiki is accurate

File: quiffen/core/qif.py [hunk 2]
```diff
@@ -184,18 +184,26 @@ def parse(cls, filepath, separator='\n', day_first=True):
                 new_account = Account.from_string(section)
                 accounts[new_account.name] = new_account
                 last_account = new_account.name
-            elif '!Type' in header_line and not accounts:
+            elif '!Type:Invst' in header_line:
+                # Investment
+                new_investment = Investment.from_string(section, separator=separator, day_first=day_first,
+                                                        line_number=line_number)
+                accounts[last_account].add_transaction(new_investment, 'Invst')
+            elif not accounts and str.lower(header_line) in VALID_TRANSACTION_ACCOUNT_TYPES:
```
Comment (isaacharrisholt) [inline quiffen/core/qif.py:192]: Can we use `header_line.lower()` here rather than the `str` classmethod?

File: quiffen/core/qif.py (hunk 1 above)
Comment (isaacharrisholt) [reply quiffen/core/qif.py]: I think this falls apart in the case where there _is_ a caret between the account definition and the transaction though, does it not?

Comment (a-bird-of-paradise) [reply quiffen/core/qif.py]: I added a test case for the circumstance where the file looks like 

```
!Type:Bank
^
D01-12-2000
T1.00
```
and it failed. I added logic which will detect when we are in a header-row-only section (which will have length two when split by `\n` due to the residue from originally splitting by `^`), and the parser now won't attempt to create a transaction when there aren't enough rows for this to happen. 

I think this kind of file is invalid per the spec but maybe there are some like it out there.

File: quiffen/core/qif.py (hunk 2 above)
Comment (a-bird-of-paradise) [reply quiffen/core/qif.py]: Now changed. I also included the whitespace stripping that is performed in other `elif` cases using `header_line.lower()`.

**Comment by isaacharrisholt** [comment]
Hi, apologies for taking so long on getting this merged. Definitely going to use this to make improvements to Quiffen, but I'm currently working on v2 so it will go into there, most likely.

I'll close this PR, but will use the fixes in v2.

**Pull request closed without merging**